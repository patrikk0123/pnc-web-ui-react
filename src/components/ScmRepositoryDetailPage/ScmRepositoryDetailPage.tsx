import { Text, TextContent, TextVariants } from '@patternfly/react-core';
import { useEffect } from 'react';
import { useParams } from 'react-router-dom';

import { scmRepositoryEntityAttributes } from 'common/scmRepositoryEntityAttributes';

import { useQueryParamsEffect } from 'hooks/useQueryParamsEffect';
import { useServiceContainer } from 'hooks/useServiceContainer';
import { useTitle } from 'hooks/useTitle';

import { ActionButton } from 'components/ActionButton/ActionButton';
import { Attributes } from 'components/Attributes/Attributes';
import { AttributesItem } from 'components/Attributes/AttributesItem';
import { BuildConfigsList } from 'components/BuildConfigsList/BuildConfigsList';
import { ContentBox } from 'components/ContentBox/ContentBox';
import { PageLayout } from 'components/PageLayout/PageLayout';
import { ScmRepositoryUrl } from 'components/ScmRepositoryUrl/ScmRepositoryUrl';
import { ServiceContainerLoading } from 'components/ServiceContainers/ServiceContainerLoading';
import { Toolbar } from 'components/Toolbar/Toolbar';
import { ToolbarItem } from 'components/Toolbar/ToolbarItem';

import * as scmRepositoryApi from 'services/scmRepositoryApi';

import { generateScmRepositoryName } from 'utils/entityNameGenerators';
import { generatePageTitle } from 'utils/titleHelper';

interface IScmRepositoryDetailPageProps {
  componentId?: string;
}

export const ScmRepositoryDetailPage = ({ componentId = 's2' }: IScmRepositoryDetailPageProps) => {
  const { scmRepositoryId } = useParams();

  const serviceContainerScmRepository = useServiceContainer(scmRepositoryApi.getScmRepository);
  const serviceContainerScmRepositoryRunner = serviceContainerScmRepository.run;

  const serviceContainerBuildConfigs = useServiceContainer(scmRepositoryApi.getBuildConfigsWithLatestBuild);
  const serviceContainerBuildConfigsRunner = serviceContainerBuildConfigs.run;

  useEffect(() => {
    serviceContainerScmRepositoryRunner({ serviceData: { id: scmRepositoryId } });
  }, [serviceContainerScmRepositoryRunner, scmRepositoryId]);

  useQueryParamsEffect(
    ({ requestConfig } = {}) => serviceContainerBuildConfigsRunner({ serviceData: { scmRepositoryId }, requestConfig }),
    { componentId }
  );
  useTitle(
    generatePageTitle({
      serviceContainer: serviceContainerScmRepository,
      entityName: generateScmRepositoryName({ scmRepository: serviceContainerScmRepository.data }),
      firstLevelEntity: 'SCM Repository',
    })
  );

  return (
    <ServiceContainerLoading {...serviceContainerScmRepository} title="SCM Repository details">
      <PageLayout
        title={generateScmRepositoryName({ scmRepository: serviceContainerScmRepository.data })}
        actions={<ActionButton link="edit">Edit SCM Repository</ActionButton>}
      >
        <ContentBox padding marginBottom isResponsive>
          <Attributes>
            <AttributesItem
              title={scmRepositoryEntityAttributes.internalUrl.title}
              tooltip={scmRepositoryEntityAttributes.internalUrl.tooltip}
            >
              {serviceContainerScmRepository.data?.internalUrl && (
                <ScmRepositoryUrl internalScmRepository={serviceContainerScmRepository.data} />
              )}
            </AttributesItem>
            <AttributesItem
              title={scmRepositoryEntityAttributes.externalUrl.title}
              tooltip={scmRepositoryEntityAttributes.externalUrl.tooltip}
            >
              {serviceContainerScmRepository.data?.externalUrl && (
                <ScmRepositoryUrl externalScmRepository={serviceContainerScmRepository.data} />
              )}
            </AttributesItem>
            <AttributesItem
              title={scmRepositoryEntityAttributes.preBuildSyncEnabled.title}
              tooltip={scmRepositoryEntityAttributes.preBuildSyncEnabled.tooltip}
            >
              {serviceContainerScmRepository.data?.preBuildSyncEnabled !== undefined &&
                (serviceContainerScmRepository.data.preBuildSyncEnabled ? 'enabled' : 'disabled')}
            </AttributesItem>
          </Attributes>
        </ContentBox>

        <Toolbar borderBottom>
          <ToolbarItem>
            <TextContent>
              <Text component={TextVariants.h2}>Usages</Text>
            </TextContent>
          </ToolbarItem>
        </Toolbar>

        <BuildConfigsList {...{ serviceContainerBuildConfigs, componentId }} />
      </PageLayout>
    </ServiceContainerLoading>
  );
};
