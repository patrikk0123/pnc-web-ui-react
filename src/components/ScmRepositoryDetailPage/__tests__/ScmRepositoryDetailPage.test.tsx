import { render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import ResizeObserver from 'resize-observer-polyfill';
import { vi } from 'vitest';

import { ScmRepositoryDetailPage } from 'components/ScmRepositoryDetailPage/ScmRepositoryDetailPage';

global.ResizeObserver = ResizeObserver;

vi.mock('services/scmRepositoryApi');
vi.mock('services/keycloakService');
vi.mock('services/uiLogger');
vi.mock('services/webConfigService');

const mockMatches = [
  { id: '0', pathname: '/', params: { scmRepositoryId: '111' }, data: undefined, handle: undefined },
  { id: '0-8', pathname: '/scm-repositories', params: { scmRepositoryId: '111' }, data: undefined, handle: 'scmRepositories' },
  {
    id: '0-8-3',
    pathname: '/scm-repositories/111',
    params: { scmRepositoryId: '111' },
    data: undefined,
    handle: undefined, //scmRepository but test is failing
  },
];

vi.mock('react-router', async () => {
  const actual = await vi.importActual<typeof import('react-router')>('react-router');
  return {
    __esModule: true,
    ...actual,
    useParams: () => ({ scmRepositoryId: '111' }),
    useMatches: () => mockMatches,
  };
});

describe('display ScmRepositoryDetailPage component', () => {
  let scmRepositoryMock: any;

  async function loadMocks() {
    const scmRepositoryRequestMock = await import('services/__mocks__/scm-repository-mock.json');
    scmRepositoryMock = scmRepositoryRequestMock;
  }

  beforeEach(async () => {
    await loadMocks();
  });

  test('renders ScmRepositoryDetail', () => {
    render(
      <MemoryRouter>
        <ScmRepositoryDetailPage />
      </MemoryRouter>
    );
  });

  test('compare snapshot with previous record', () => {
    const tree = render(
      <MemoryRouter>
        <ScmRepositoryDetailPage />
      </MemoryRouter>
    );
    expect(tree.container).toMatchSnapshot();
  });

  test('renders ScmRepositoryDetail to have the right data', async () => {
    render(
      <MemoryRouter>
        <ScmRepositoryDetailPage />
      </MemoryRouter>
    );
    await waitFor(() => {
      const displayName = screen.getByText(scmRepositoryMock.displayName);
      expect(displayName).toBeInTheDocument();
    });
  });
});
