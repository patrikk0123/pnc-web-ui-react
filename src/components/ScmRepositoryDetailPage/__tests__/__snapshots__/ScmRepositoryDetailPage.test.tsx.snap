// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`display ScmRepositoryDetailPage component > compare snapshot with previous record 1`] = `<div />`;
